### PR TITLE
Restart daemon after it dies

### PR DESCRIPTION
**Pull request by brummett**
If the socket to the daemon process is gone when we try to make a request, restart the daemon

### CODE REVIEW

File: t/Nessy/daemon_dies_after_request.t [hunk 1]
```diff
@@ -11,31 +14,70 @@ use AnyEvent;
...
+    alarm(0);
+
+    ok(! $claim, "couldn't make a claim when the daemon closed its socket");
+    ok(!$@, 'no exception');
+    like($warning_message,
+        qr(claim $claim_name at .* failed: Connection reset by peer),
+        'warning message');
+};
+
+subtest 'daemon restarts after going away' => sub {
```
Comment (nnutter) [inline t/Nessy/daemon_dies_after_request.t]: I'm not seeing where the `socket_watcher` would get undefined to trigger another `_fork_and_run_daemon`.  Should we add to the test to observer that `_fork_and_run_daemon` was run a second time?

File: lib/Nessy/Client.pm [hunk 2]
```diff
@@ -89,15 +106,17 @@ sub _close_unused_socket {
...
 sub _run_child_process {
-    my($class, %params) = @_;
+    my($self, @sockets) = @_;
+
+    my $params = $self->constructor_params();
 
     eval {
-        $class->_close_unused_socket($params{socketpair}->[PARENT_PROCESS_SOCKET]);
-        my $daemon_class = $class->_daemon_class_name;
+        $self->_close_unused_socket($sockets[PARENT_PROCESS_SOCKET]);
```
Comment (nnutter) [inline lib/Nessy/Client.pm]: It's not really part of this PR but I liked the naming convention you used in the test: `daemon_to_client_sock` and `client_to_daemon_sock`.  It might also be an improvement to adopt `client` instead of `parent` and `daemon` instead of `child`.

**Comment by nnutter** [comment]
I moved my messages from the commits to _Files changed_ so you probably got confusing double notifications.

File: t/Nessy/daemon_dies_after_request.t (hunk 1 above)
Comment (brummett) [inline t/Nessy/daemon_dies_after_request.t]: ```socket_watcher``` gets cleared in ```bailout()```

The second subtest hooks into fork() to ensure it actually forks as part of ```_fork_and_run_daemon()```.  It doesn't actually fork again until we try to make another lock.

File: lib/Nessy/Client.pm (hunk 2 above)
Comment (brummett) [inline lib/Nessy/Client.pm]: Neither naming convention is perfect, but the one in the test is less bad, I think.  I'll change it.

File: t/Nessy/daemon_dies_after_request.t (hunk 1 above)
Comment (nnutter) [inline t/Nessy/daemon_dies_after_request.t]: Ah, ok, I see it now.

File: t/Nessy/daemon_dies_after_request.t [hunk 3]
```diff
@@ -11,31 +14,70 @@ use AnyEvent;
...
+    plan tests => 5;
+
+    # when it gets the claim message, send it back as successful
+    my $respond_successfully = sub {
+                my($watcher, $message_hash) = @_;
+                my $message = Nessy::Client::Message->new(%$message_hash);
+                ok($message, 'Got message from client to make a lock');
```
Comment (nnutter) [inline t/Nessy/daemon_dies_after_request.t]: I think the test name should probably just be "received a message" because I think it's trigger for both `claim` and `release` below.

**Comment by brummett** [comment]
Changed the test message and renamed the constants

**Comment by nnutter** [comment]
:+1:
